File: al_toolbox/week1/max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>


using std::cin;
using std::cout;
using std::endl;
using std::vector;

int64_t max_pairwise_product(vector<int>& numbers){
	int64_t max_product = 0;
	int n = numbers.size();

	for (int first = 0; first < n; ++first){
		for (int second = first + 1; second < n; ++second){
			max_product = std::max(
					max_product, 
					int64_t(numbers[first]) * numbers[second]
                    );
		}
	}
	return max_product;
}



int64_t max_pairwise_product_fast(vector<int>& numbers){
	int n = numbers.size();

    int max_index1 = 0;
	for (int i = 0; i < n; ++i)
        if (numbers[i] > numbers[max_index1])
            max_index1 = i;
   
    int max_index2 = -1;
	for (int i = 0; i < n; ++i)
        if (i != max_index1 
                && (max_index2 == -1 || numbers[i] > numbers[max_index2]))
            max_index2 = i;

	return int64_t(numbers[max_index1]) * numbers[max_index2];
}



int main(){
    int n;
	cin >> n;
	vector<int> input_numbers(n);
	for (int i = 0; i < n; i++)
		cin >> input_numbers[i];

	cout << max_pairwise_product_fast(input_numbers) << endl; 
    return 0;
}

File: al_toolbox/week1/max_pairwise_product.py
#!/usr/bin/env python3


def max_pairwise_product(numbers, n):
    max_product = 0
    for i in range(n):
        for j in range(i + 1, n):
            max_product = max(
                max_product,
                numbers[i] * numbers[j])
    return max_product


def max_pairwise_product_fast(numbers, n):
    max_index1 = 0
    for i in range(n):
        if numbers[i] > numbers[max_index1]:
            max_index1 = i

    max_index2 = -1
    for i in range(n):
        if i != max_index1 \
                and (max_index2 == -1 or numbers[i] > numbers[max_index2]):
            print(i)
            max_index2 = i

    return numbers[max_index1] * numbers[max_index2]


if __name__ == "__main__":
    n = int(input())
    input_numbers = [int(x) for x in input().split()]
    print(max_pairwise_product_fast(input_numbers, n))
